### PR TITLE
Add support for Windows 10 Native SSH agent #5

### PR DESCRIPTION
**Pull request by ngyuki**

### CODE REVIEW

File: sshagent_windows.go [hunk 1]
```diff
@@ -22,22 +22,46 @@
...
-
-	return agent.NewClient(&conn{}), nil, nil
+	conn, err := winio.DialPipe(sshAgentPipe, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf(
+			"SSH agent requested but Pageant not running and Error %v",
+			err,
+		)
+	}
+	return agent.NewClient(conn), nil, nil
```
Comment (svanharmelen) [inline sshagent_windows.go:64]: I wonder if we should make the Windows Native agent the default (i.e. test that one first). What do you think? Also I think we need to make the error more clear in that it tested for both Pageant and the native SSH agent and that neither are available.

**Review by svanharmelen — COMMENTED** [review]
@ngyuki I'm tempted to merge this PR, but I no longer have access to a Windows machine to validate these changes. So can you confirm that you really tested all three options? So a standard SSH agent on Linux, Pageant and the native SSH agent on Windows?

File: sshagent_windows.go (hunk 1 above)
Comment (ngyuki) [reply sshagent_windows.go:64]: > I wonder if we should make the Windows Native agent the default (i.e. test that one first). What do you think?

Considering that only Pageant was supported originally, I think the compatibility is better if you make Pageant the default.

> Also I think we need to make the error more clear in that it tested for both Pageant and the native SSH agent and that neither are available.

For example, like following?

```
SSH agent requested but Pageant not running and cannot connect native SSH agent.
```

Comment (svanharmelen) [reply sshagent_windows.go:64]: > Considering that only Pageant was supported originally, I think the compatibility is better if you make Pageant the default.

Agree...

Or something like:
```
SSH agent requested, but could not detect Pageant or Windows native SSH agent.
```

**Comment by ngyuki** [comment]
I use following simple code to confirmed in three cases.

- ssh-agent (Fedora 31 on WSL1 / OpenSSH_8.1p1)
- pageant.exe (PuTTy 0.74)
- native ssh agent on Windows

```go
package main

import (
	"os"

	sshagent "github.com/xanzy/ssh-agent"
	"golang.org/x/crypto/ssh"
)

const (
	addr = "192.0.2.123:22"
	user = "ngyuki"
)

func main() {
	agentClient, _, err := sshagent.New()
	if err != nil {
		panic(err)
	}

	config := &ssh.ClientConfig{
		User: user,
		Auth: []ssh.AuthMethod{
			ssh.PublicKeysCallback(agentClient.Signers),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
	}

	client, err := ssh.Dial("tcp", addr, config)
	if err != nil {
		panic(err)
	}

	session, err := client.NewSession()
	if err != nil {
		panic(err)
	}
	defer session.Close()

	session.Stdout = os.Stdout
	session.Stderr = os.Stderr

	err = session.Run("uname -a")
	if err != nil {
		panic(err)
	}
}
```

go.mod is as follows.

```go
module github.com/ngyuki/ssh-agentt-test

go 1.13

require (
	github.com/Microsoft/go-winio v0.4.14
	github.com/xanzy/ssh-agent v0.2.1
	golang.org/x/crypto v0.0.0-20200820211705-5c72a883971a
)

replace github.com/xanzy/ssh-agent v0.2.1 => github.com/ngyuki/ssh-agent v0.2.2-0.20200828065058-75dc0d7786e1
```

**Review by svanharmelen — APPROVED** [review]
Thank you very much for the PR and for testing it properly @ngyuki! Very much appreciated!

**Comment by kstkn** [comment]
@svanharmelen could you also please create a release for this 🙏

**Comment by svanharmelen** [comment]
Done 👍🏻

**Pull request merged**